File: day09/day09p2.py
import collections
from dataclasses import dataclass
import sys
from typing import Optional


MAX_SIZE = 9


@dataclass
class Block:
    index: int
    size: int
    f_id: int


nums = [int(n) for n in sys.stdin.readlines()[0].rstrip()]

# block size -> list of indices
free_blocks: collections.defaultdict[int, list[int]] = collections.defaultdict(list)

counter = 0
fs: list[Block] = []
free_flag = False
index = 0
for n in nums:
    if free_flag:
        if n:
            free_blocks[n].append(index)
    else:
        fs.append(Block(index=index, size=n, f_id=counter))
        counter += 1
    index += n
    free_flag = not free_flag


def find_first_free(
    size: int, free_blocks: dict[int, list[int]]
) -> Optional[tuple[int, int]]:
    options = []
    for s in range(size, MAX_SIZE + 1):
        if free_blocks[s]:
            options.append((s, min(free_blocks[s])))
    return min(options, key=lambda t: t[1]) if options else None


for block in reversed(fs):
    find_result = find_first_free(block.size, free_blocks)
    if find_result is None:
        continue
    free_block_size, free_block_index = find_result
    if free_block_index > block.index:
        continue
    block.index = free_block_index
    free_blocks[free_block_size].remove(free_block_index)
    new_free_size = free_block_size - block.size
    new_free_index = free_block_index + block.size
    if new_free_size:
        free_blocks[new_free_size].append(new_free_index)

total = 0
for b in fs:
    total += sum(b.f_id * n for n in range(b.index, b.index + b.size))
print(total)

File: day09/day09p2.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <map>
#include <queue>

namespace {
constexpr unsigned int MAX_SIZE = 9;

struct Block {
  std::size_t index;
  unsigned int size;
  std::size_t f_id;
};

using FreeBlocks =
    std::map<int, std::priority_queue<std::size_t, std::vector<std::size_t>,
                                      std::greater<std::size_t>>>;

// Returns size of the first block that can fit `size` bytes or 0 if none exist.
unsigned int find_first_free(unsigned int size, FreeBlocks& free_blocks) {
  std::vector<std::pair<unsigned int, std::size_t>> options;
  unsigned int min_size = std::numeric_limits<unsigned int>::max();
  std::size_t min_index = std::numeric_limits<std::size_t>::max();
  // std::map has keys in sorted order
  for (auto it = free_blocks.find(size); it != free_blocks.end(); ++it) {
    if (!it->second.empty()) {
      auto index = it->second.top();
      if (index < min_index) {
        min_size = it->first;
        min_index = index;
      }
    }
  }
  if (min_index != std::numeric_limits<std::size_t>::max()) {
    return min_size;
  }
  return 0;
}

}  // namespace

int main() {
  FreeBlocks free_blocks;
  for (std::size_t i = 0; i <= MAX_SIZE; ++i) {
    free_blocks[i];
  }
  std::vector<Block> fs{};

  char digit{};
  std::size_t counter{};
  bool free_flag = false;
  std::size_t index{};
  while (std::cin >> digit) {
    unsigned int n = digit - '0';
    if (free_flag) {
      if (n) {
        free_blocks[n].push(index);
      }
    } else {
      fs.emplace_back(index, n, counter);
      ++counter;
    }
    index += n;
    free_flag = !free_flag;
  }

  for (auto it = fs.rbegin(); it != fs.rend(); ++it) {
    auto& block = *it;
    auto free_block_size = find_first_free(block.size, free_blocks);
    if (free_block_size == 0) {
      continue;
    }
    auto free_block_index = free_blocks[free_block_size].top();
    if (free_block_index > block.index) {
      continue;
    }
    block.index = free_block_index;
    auto new_free_size = free_block_size - block.size;
    auto new_free_index = free_block_index + block.size;
    free_blocks[free_block_size].pop();
    if (new_free_size) {
      free_blocks[new_free_size].push(new_free_index);
    }
  }

  std::size_t total{};
  for (const auto& b : fs) {
    auto low = b.index;
    auto high = b.index + b.size - 1;
    // sum of range
    total += b.f_id * (((low + high) * (high - low + 1)) / 2);
  }
  std::cout << total << '\n';
}
